### PR TITLE
i2c drivers

### PR DESCRIPTION
**Pull request by Sabramz**
These are probably ass. Also the includes aren't there cuz idk what i need to include. Just tell me and ill add em.

### CODE REVIEW

File: driver/include/i2c.hpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/**
+ * @file include/i2c.hpp
+ * @brief Drivers for I2C using the kernel smbus driver.
+ */
+#pragma once
+
```
Comment (vladaviedov) [inline driver/include/i2c.hpp:6]: cstdint missing

File: driver/include/i2c.hpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+class i2c {
+
+public:
+
+    /**
+     * @brief Constructor. Initializes I2C device.
+     *
+     * @param[in] dev_addr - address of the I2C slave device.
+     */
+    i2c(const uint8_t dev_addr);
```
Comment (vladaviedov) [inline driver/include/i2c.hpp]: Argument and member name are the same

File: driver/include/i2c.hpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    /** 
+     * @brief I2C read.
+     *
+     * @param[in] data - Data buffer to write to.
```
Comment (vladaviedov) [inline driver/include/i2c.hpp]: Data buffer is out, if you are reading

File: driver/include/i2c.hpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    /** 
+     * @brief I2C write.
+     *
+     * @param[in] reg - Register to read from.
+     * @param[in] data - Data buffer to write.
+     * @param[in] size - Size of data buffer.
+     * @return error code
+     */
+    int write(uint8_t reg, char* data, uint8_t size);
```
Comment (vladaviedov) [inline driver/include/i2c.hpp]: size - make uint32_t
data - const
return val - make boolean

File: driver/src/i2c.cpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ * @file src/i2c.cpp
+ * @brief I2C driver.
+ */
+#include "i2c.hpp"
+
+#include <linux/i2c-dev.h>
+#include <i2c/smbus.h>
+
+i2c::i2c(uint8_t dev_addr) {
+    int fd;
```
Comment (vladaviedov) [inline driver/src/i2c.cpp]: remove, fd is a member

File: driver/src/i2c.cpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    int adapter_nr = 1;
```
Comment (vladaviedov) [inline driver/src/i2c.cpp]: Make it a constexpr in device.hpp
(But also pass it as an argument to the constructor)

File: driver/src/i2c.cpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    char filename[32];
+
+    snprintf(filename, 31, "/dev/i2c-%d", adapter_nr);
```
Comment (vladaviedov) [inline driver/src/i2c.cpp]: 32, snprintf accounts for null char; std::snprintf in cpp

File: driver/src/i2c.cpp [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        throw std::runtime_error("Failed to open i2c port");
+    }
+
+    if (ioctl(fd, I2C_SLAVE, addr) < 0) {
+        throw std::runtime_error("Failed to open i2c port (ioctl)");
+    }
+}
+
+i2c::read(char* data, uint8_t size) {
+    size += 1;
```
Comment (vladaviedov) [inline driver/src/i2c.cpp]: This will cause a segfault

File: driver/src/i2c.cpp [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    if (fd < 0) {
+        throw std::runtime_error("Failed to open i2c port");
+    }
+
+    if (ioctl(fd, I2C_SLAVE, addr) < 0) {
+        throw std::runtime_error("Failed to open i2c port (ioctl)");
+    }
+}
+
+i2c::read(char* data, uint8_t size) {
```
Comment (vladaviedov) [inline driver/src/i2c.cpp]: Missing return type

File: driver/include/i2c.hpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    i2c(const uint8_t dev_addr);
+
+    /** 
+     * @brief I2C read.
+     *
+     * @param[in] data - Data buffer to write to.
+     * @param[in] size - Size of data buffer.
+     * @return error code
+     */
+    int read(char* data, uint8_t size);
```
Comment (vladaviedov) [inline driver/include/i2c.hpp]: Return the size of the read buffer

read/write functions should be marked const

File: driver/src/i2c.cpp [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    size += 1;
+    if (read(fd, data, size) != size) {
+        throw std::runtime_error("Failed to read i2c");
+    }
+}
+
+i2c::write(uint8_t reg, char* data, uint8_t size) {
+    char* buf;
+    *buf[0] = reg;
```
Comment (vladaviedov) [inline driver/src/i2c.cpp]: Segfault

File: driver/src/i2c.cpp [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    std::copy(data, data + size, buf + 1);
+    // size is now the size of buf
+    size += 1;
+    if (write(fd, data, size) != size) {
+        throw std::runtime_error("Failed to write i2c");
```
Comment (vladaviedov) [inline driver/src/i2c.cpp]: Return int result

File: .gitignore [hunk 13]
```diff
@@ -3,3 +3,6 @@ build/
 # OS specific stuff
 .DS_Store
 *~
+
+# Docker
+Dockerfile
```
Comment (vladaviedov) [inline .gitignore]: It's fine to commit the dockerfile if you have a working config

File: driver/include/i2c.hpp [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+public:
+
+    /**
+     * @brief Constructor. Initializes I2C device.
+     *
+     * @param[in] dev_addr - address of the I2C slave device.
+     */
+    i2c(const uint8_t dev_addr);
+
```
Comment (vladaviedov) [inline driver/include/i2c.hpp:19]: Missing destructor

**Review by vladaviedov — CHANGES_REQUESTED** [review]
See my comments.
Please put newlines at the ends of files.
We need i2c-tools to use SMBus, I will add that today.

File: driver/include/i2c.hpp (hunk 4 above)
Comment (Sabramz) [reply driver/include/i2c.hpp]: Why boolean?

Comment (vladaviedov) [reply driver/include/i2c.hpp]: It should show if write was successful or not, returning bytes written isn't very helpful above this layer

Comment (Sabramz) [reply driver/include/i2c.hpp]: ok ez

**Review by vladaviedov — APPROVED** [review]

**Pull request merged**